plugins/inputs/pf/pf.go

```diff
@@ -0,0 +1,207 @@
+}
+
+var stateTableHeaderRE = regexp.MustCompile("^State Table")
+var anyTableHeaderRE = regexp.MustCompile("^[A-Z]")
+
+func (pf *PF) parsePfctlOutput(pfoutput string, acc telegraf.Accumulator) error {
+	lines := strings.Split(pfoutput, "\n")
+	stateTableFound := false
+	for i, line := range lines {
+		if stateTableHeaderRE.MatchString(line) {
+			endline := len(lines)
+			for j, el := range lines[i+1:] {
+				if anyTableHeaderRE.MatchString(el) {
+					endline = j + i + 1
+					break
+				}
+			}
+			if perr := pf.parseStateTable(lines[i+1:endline], acc); perr != nil {
+				return perr
+			}
+			stateTableFound = true
+		}
+	}
+	if !stateTableFound {
+		return errParseHeader
+	}
+	return nil
+}
+
+var stateTableRE = regexp.MustCompile(`^  (.*?)\s+(\d+)`)
+
+func (pf *PF) parseStateTable(lines []string, acc telegraf.Accumulator) error {
+	st := StateTable{}
+	tags, err := st.getTags()
+	if err != nil {
+		return fmt.Errorf("Can't retrieve struct tags: %v", err)
+	}
+	fMap := make(map[string]bool)
+	for i := 0; i < len(tags); i++ {
+		fMap[tags[i]] = false
+	}
+
+	for _, v := range lines {
+		entries := stateTableRE.FindStringSubmatch(v)
+		if entries != nil {
+			fs, err := st.setByTag(entries[1], entries[2])
+			if err != nil {
+				return errors.New("can't set statetable field from tag")
+			}
+			if fs {
+				fMap[entries[1]] = true
+			}
+		}
+	}
+
+	for k, v := range fMap {
+		if !v {
+			return errMissingData(k)
+		}
+	}
+
+	fields := make(map[string]interface{})
+	fields["entries"] = st.CurrentEntries
+	fields["searches"] = st.Searches
+	fields["inserts"] = st.Inserts
+	fields["removals"] = st.Removals
+	acc.AddFields(measurement, fields, make(map[string]string))
+	return nil
+}
+
+func (pf *PF) callPfctl() (string, error) {
+	c, err := pf.buildPfctlCmd()
+	if err != nil {
+		return "", fmt.Errorf("Can't construct commandline: %s", err)
+	}
+	out, oerr := c.Output()
+	if oerr != nil {
+		return string(out), fmt.Errorf("error running %s: %s: %s", pfctlCommand, oerr, oerr.(*exec.ExitError).Stderr)
+	}
+	return string(out), oerr
+}
+
+var execLookPath = exec.LookPath
+var execCommand = exec.Command
+
+func (pf *PF) buildPfctlCmd() (*exec.Cmd, error) {
+	cmd, err := execLookPath(pfctlCommand)
+	if err != nil {
+		return nil, fmt.Errorf("can't locate %s: %v", pfctlCommand, err)
+	}
+	args := []string{"-s", "info"}
+	if pf.UseSudo {
+		args = append([]string{cmd}, args...)
+		cmd, err = execLookPath("sudo")
+		if err != nil {
+			return nil, fmt.Errorf("can't locate sudo: %v", err)
+		}
+	}
+	c := execCommand(cmd, args...)
+	return c, nil
+}
+
+type StateTable struct {
+	CurrentEntries uint32 `pfctl:"current entries"`
+	Searches       uint64 `pfctl:"searches"`
+	Inserts        uint64 `pfctl:"inserts"`
+	Removals       uint64 `pfctl:"removals"`
+}
+
+func (pf *StateTable) getTags() ([]string, error) {
+	tags := []string{}
+	structVal := reflect.ValueOf(pf).Elem()
```

Comment:
I don't think it is worth using reflect in this case, reflect is slow, hard to understand, and there are not so many fields that we can't just do it directly.